content/blog/2019-12-18-wasm-to-evm.md

```diff
@@ -0,0 +1,83 @@
+
+One approach is to develop an LLVM backend to EVM. This project is going to investigate another approach by binary translation. 
+
+## Binary Translator 
+
+Since the Ewasm standard is not finalized yet, there is no existing compiler for Ewasm, so this proof-of-concept project will focus on translating a single computation-based function from Wasm to EVM. In the following discussion, we only consider the bytecode of the function and ignore other wrapping code.
+
+For example, given a simple C function "add", 
+```C
+int add(int a, int b) {
+    return a + b;
+}
+```
+it can be compiled to Wasm bytecode `200120006a`.
+```
+...
+20    ; get_local
+01    ; $01
+20    ; get_local
+00    ; $00
+6a    ; i32.add
+...
+```
+Wasm has a stack-based design, so the code means
+
+1. push parameter $1 onto the stack
+2. push parameter $0 onto the stack
+3. pop the top two elements from the stack, add them, and push the result back to the stack
+
+This can be translated to EVM bytecode `60243520042501`
+
+```
+...
+6024    ; PUSH 0x04
+35      ; CALLDATALOAD load second parameter at 0x04
+6004    ; PUSH 0x24
+35      ; CALLDATALOAD load first parameter at 0x24
+01      ; ADD add top two elements on stack
+...
+```
+
+Since both architectures are stack-based, the implementation of the basic binary translator is straight-forward. Many logic and arithmetic operations can be translated directly according to the opcode table. Here I will discuss some interesting optimizations instead. 
+
+In EVM, there is a concept called "gas", which is a metering unit for measuring the cost of each instruction. The gas price is calculated by electricity and storage used for running the instruction. The smart contract users need to pay the price to the miners who run the smart contract application. 
+Unfortunately, the gas price for Ewasm instructions is not out yet, we cannot compare the total cost for running an Ewasm program and a translated EVM program, but we can still optimize EVM bytecode to reduce the gas price of a function.
+
+In Wasm, function parameters are passed by local variables. They are accessed by `get_local` instruction followed by an index number. Similarly, function parameters are stored in memory, which can be accessed by memory load operation `MLOAD` with an offset. For example, in the above function with two uint256 parameters, the first parameter is at 0x04, the second parameter is at 0x24. 
+
+Wasm compiler tends to use the local variable like registers, the compiled bytecode will load data by `get_local` whenever it needs it. However, this approach is not optimal for EVM, especially when the program needs to load a parameter multiple times in a short period. 
```

Comment:
- *The* Wasm compiler tends, or *A* Wasm compiler tends, or Wasm *compilers* *tend*
- start a new sentence at "the compiled bytecode"